torch/csrc/jit/passes/batch_mm.cpp

```diff
@@ -68,52 +70,98 @@ namespace torch { namespace jit {
 // the trees we formed and fuse them.
 
 // Tunable parameter. Set to something larger if it turns out to be better.
-static constexpr size_t min_fusion_size = 2;
+static constexpr size_t min_fusion_size = 4;
 
-static std::array<int64_t, 2> as_array(at::IntList sizes) {
-  JIT_ASSERT(sizes.size() == 2);
-  std::array<int64_t, 2> arr = {sizes[0], sizes[1]};
-  return arr;
+bool have_same_shape(at::TensorList inputs) {
+  auto expected_sizes = inputs[0].sizes();
+  return std::all_of(inputs.begin(), inputs.end(),
+                     [expected_sizes](const at::Tensor& t) {
+                       return t.sizes() == expected_sizes;
+                     });
 }
 
+bool shape_is_fast(const at::Tensor& lhs, const at::Tensor& rhs) {
+  size_t l = lhs.size(0);
+  size_t m = lhs.size(1);
+  size_t r = rhs.size(1);
+  // Numbers obtained by some simple benchmarks of fp32 gemms on a TITAN V
+  return m < 512 || ((m < 256 && r < 256) || (m > 256 && r > 256));
+}
+
+RegisterOperators mm_tree_reduction_reg({
+  Operator(
+    Symbol::prim("MMTreeReduce"),
```

Comment:
This doesn't work with autodiff, right? So this optimization would create a MMTreeReduce for an LSTM backward, but the double backward could possibly benefit from this being autodiff-able.